File: Exercises2024/Chapter04_basic_mc/C/3-HardDisks/dynamic/Source/hard2.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include "system.h"
#include "ran_uniform.h"

#define CycleMultiplication 1000

int PBC;
// hard disks on a square; random placement

int main(void)
{
  int NumberOfCycles,NumberOfInitializationCycles;
  int i,j,k,SelectedParticle,NumberOfSamples,overlap;
  int NumberOfAttemptedMoves,NumberOfAcceptedMoves;
  double r2,MaximumDisplacement;
  VECTOR PosNew,dr;
  FILE *FilePtr;
      
  // initialize the random number generator with the system time
  InitializeRandomNumberGenerator(time(0l));

  FilePtr=fopen("movie.pdb","w");

  // read the input parameters
  printf("How many cycles (x %d)      ? ",CycleMultiplication);
  fscanf(stdin,"%d",&NumberOfCycles);

  printf("How many init. cycles ?        (example: 100      ) ");
  fscanf(stdin,"%d",&NumberOfInitializationCycles);


  printf("How many particles    ?        (always 2 < i < 80 ) ");
  fscanf(stdin,"%d",&NumberOfParticles);

  printf("Maximum displacement  ?        (disp > 0          ) ");
  fscanf(stdin,"%lf",&MaximumDisplacement);

  printf("Period boundary conditions ?   (0 or 1) ");
  fscanf(stdin,"%d",&PBC);

  if((NumberOfParticles>MAX_PARTICLES)||(NumberOfParticles<2))
  {
    printf("input error\n");
    exit(0);
  }

  Sample(INITIALIZE);

  NumberOfAttemptedMoves=0;
  NumberOfAcceptedMoves=0.0;
  NumberOfSamples=0;

  // put particles on a lattice 

  k=0;
  for(i=0;i<(int)(BOXSIZE-1.0);i++)
  {
    for(j=0;j<(int)(BOXSIZE-1.0);j++)
    {
      Positions[k].x=i*BOXSIZE/(BOXSIZE-1.0);
      Positions[k].y=j*BOXSIZE/(BOXSIZE-1.0);
      k++;
    }
  }

  for(i=0;i<NumberOfCycles;i++)
  {
    for(j=0;j<CycleMultiplication;j++)
    {
      NumberOfAttemptedMoves++;

      // select particle at random 
      SelectedParticle=RandomNumber()*NumberOfParticles;

      PosNew.x=Positions[SelectedParticle].x+(RandomNumber()-0.5)*MaximumDisplacement;
      PosNew.y=Positions[SelectedParticle].y+(RandomNumber()-0.5)*MaximumDisplacement;

      // always reject when particle is out of the box
      //if(PBC||(MIN(PosNew.x,PosNew.y)>=0.0&&MAX(PosNew.x,PosNew.y)<BOXSIZE))
      if(PBC||((!PBC)&&(MIN(PosNew.x,PosNew.y)>=0.0&&MAX(PosNew.x,PosNew.y)<BOXSIZE)))
      {
        // see if there is an overlap with any other particle
        overlap=FALSE;
        for(k=0;((k<NumberOfParticles));k++)
        {
          if(k!=SelectedParticle)
          {
            dr.x=PosNew.x-Positions[k].x;
            dr.y=PosNew.y-Positions[k].y;

            // apply boundary conditions
            if(PBC)
            {
              dr.x-=BOXSIZE*rint(dr.x/BOXSIZE);
              dr.y-=BOXSIZE*rint(dr.y/BOXSIZE);
            }

            r2=SQR(dr.x)+SQR(dr.y);
            if(r2<1.0) overlap=TRUE;
          }
        }
        if(overlap==FALSE)
        {
          NumberOfAcceptedMoves++;
          Positions[SelectedParticle].x=PosNew.x;
          Positions[SelectedParticle].y=PosNew.y;
        }
      }
      if(((j%5)==0)&&(i>NumberOfInitializationCycles))
      {
        Sample(SAMPLE);
        NumberOfSamples++;
      }
    }
    if((i%5)==0) WritePdb(FilePtr);
  }

  if(NumberOfSamples>4) Sample(WRITE_RESULTS);
  printf("Fraction Succes : %f\n",(double)NumberOfAcceptedMoves/NumberOfAttemptedMoves);

  fclose(FilePtr);
  return 0;
}

File: Exercises2024/Chapter04_basic_mc/Fortran/3-HardDisks/dynamic/Source/hard2.f
      Program Hard2
      Implicit None

      Include 'system.inc'

Ccccccccccccccccccccccccccccccccccccccccccccccccc
C     Hard Disks On A Square; Random Placement  C
Ccccccccccccccccccccccccccccccccccccccccccccccccc

      Integer Nstep,I,J,n,k,Ipart,Pp,
     &     Ninit,Sstmm,CycleMultiplication,Overlap
      Double Precision Ran_Uniform,R2,NAttempted,NAccepted,Xn,Yn,
     &     Dispmax,M1,DX,DY

      Parameter (CycleMultiplication = 1000)
      
Cccccccccccccccccccccccccccccccccccccccccccc
C     Initialize Random Number Generator   C
Cccccccccccccccccccccccccccccccccccccccccccc

      M1 = 0.001d0*Dble(Mod((10+10*Sstmm()),1000))

      If(M1.Lt.0.001d0) M1 = 0.001d0
      If(M1.Gt.0.999d0) M1 = 0.999d0

      Call Genrand(M1)

Cccccccccccccccccccccccccccccccccccccccccccc
C     Read Info From Disk                  C
Cccccccccccccccccccccccccccccccccccccccccccc

      Write(*,*) 'How Many Cycles (x ',CycleMultiplication,') ?)'
      Read(*,*)  Nstep

      Write(*,*) 'How Many Init. Cycles ?        (Example: 100      )'
      Read(*,*)  Ninit
      
      Write(*,*) 'How Many Particles    ?        (Always 2 < I < 80 )'
      Read(*,*)  Npart

      Write(*,*) 'Maximum Displacement  ?        (Disp > 0          )'
      Read(*,*)  Dispmax

      Write(*,*) 'Periodic boundary conditions ? (0 or 1)'
      Read(*,*) PBC

      If(Npart.Gt.80.Or.Npart.Lt.2) Stop

      Call Sample(1)

      NAttempted = 0.0d0
      NAccepted = 0.0d0
      Pp   = 0

Ccccccccccccccccccccccccccccccccccc
C     Put Particles On A Lattice  C
Ccccccccccccccccccccccccccccccccccc

      k = 0

      Do I=1,9
         Do J=1,9

            k    = k + 1
            X(k) = Dble(I)*1.1d0
            Y(k) = Dble(J)*1.1d0
         Enddo
      Enddo

      Do n=1,Nstep
         Do k=1,CycleMultiplication

            NAttempted = NAttempted + 1.0d0 

Ccccccccccccccccccccccccccccccccccc
C     Select Particle At Random   C
Ccccccccccccccccccccccccccccccccccc

            Ipart = 1        + Int(Ran_Uniform()*Dble(Npart))
            Xn    = X(Ipart) + (Ran_Uniform()-0.5d0)*Dispmax
            Yn    = Y(Ipart) + (Ran_Uniform()-0.5d0)*Dispmax

Ccccccccccccccccccccccccccccccccccccccccccccccccccccccc
C     Always Reject When Particle Is Out Of The Box   C
Ccccccccccccccccccccccccccccccccccccccccccccccccccccccc
            
            If((PBC.eq.1).or.((PBC.eq.0).and.(Min(Xn,Yn).Ge.0.0d0)
     &         .and.(Max(Xn,Yn).Lt.BoxSize))) then
          
               Overlap=0

Ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc
C     See If There Is An Overlap With Any Other Particle  C
Ccccccccccccccccccccccccccccccccccccccccccccccccccccccccccc

               Do I=1,Npart
                  If(I.Ne.Ipart) Then
  
                     if(PBC.eq.0) then
                       R2 = (Xn-X(I))**2 + (Yn-Y(I))**2
                     else
                       DX=Xn-X(I)
                       DX=DX-BoxSize*DNINT(DX/BoxSize)
                       DY=Yn-Y(I)
                       DY=DY-BoxSize*DNINT(DY/BoxSize)
                       R2 = DX**2 + DY**2
                     endif

                     If(R2.Lt.1.0d0) Overlap=1
                  Endif
               Enddo

Cccccccccccccccccccccccccccccccccc
C     No Overlaps, So Accepted   C
Cccccccccccccccccccccccccccccccccc

               If(Overlap.eq.0) then
                  NAccepted    = NAccepted + 1.0d0
                  X(Ipart) = Xn
                  Y(Ipart) = Yn
               Endif
            Endif

            If(Mod(k,5).Eq.0.And.n.Gt.Ninit) Then
               Call Sample(2)
               Pp = Pp + 1
            Endif
         Enddo
         If(Mod(n,5).Eq.0) Call Writepdb
      Enddo

      If(Pp.Gt.4) Call Sample(3)
      Write(6,*) 'Fraction Success : ',NAccepted/NAttempted

      Stop
      End
